README
RNAHeliCes
Copyright (C) 2011-20 Jiabin Huang, Björn Voss.
Send comments/bug reports to: J. Huang <j.huang@uke.de>.


Quickstart
------------
1. Read deps and install the dependent compilers and packages
2. Type
     ./configure
     make
     make install
   to configure, make and install RNAHeliCes, HiPath and HiTed
   Notes:
     - If there are any linking problems after installing, please check if /usr/local/lib is contained in the environmental variable LD_LIBRARY_PATH. 
     - ./src/libs/libRNA.a contains all Vienna package routines, this may be system dependent. If so, please reload this from Vienna RNA Package.
3. Type './src/RNAHeliCes ./examples/collosoma_slrna.seq' to run an example.
4. Type './src/HiPath -h' to get help for calculating energy barriers of an energy landscape.
   Type './src/HiPath -f examples/switches_4.faa -k 40 -P ./librna/vienna/rna_turner1999.par' to run an example.
5. Type './src/HiTed -h' to get help for calculating minimum Hishape based Tree edit distance.
   Type './src/HiTed ./examples/riboswitches.fas -t 1 -r 1' to run an example.
6. See ./HiKinetics/README for HiKinetics.

See 
    "INSTALL"        for detailed installation instructions, and
    "COPYING"        for disclaimer and copyright.
    
 
Citations:
Huang, Jiabin & Voss, Björn. (2014). Analysing RNA-kinetics based on folding space abstraction. BMC Bioinformatics 15(1):60. DOI: 10.1186/1471-2105-15-60
Huang, Jiabin & Backofen, Rolf & Voss, Björn. (2012). Abstract folding space analysis based on helices. RNA (New York, N.Y.). 18. DOI: 10.1261/rna.033548.112. 
Huang, Jiabin & Voss, Björn. (2011). RNAHeliCes – Folding space analysis based on position aware structure abstraction. Conference: Proceedings of the German Conference on Bioinformatics 2011 (GCB 2011)At: RT--31



ChangeLog
Version 2.0.14   Jan. 14, 2014
        * add option for theta and format description

Version 2.0.13   Dec. 10, 2013
        * add a front-end for HiKinetics

Version 2.0.12   Feb. 09, 2013
        * minor

Version 2.0.11   Jan. 31, 2013
        * add an option -P for assigning partition functions of hishape classes

Version 2.0.10   Dec. 20, 2012
        * use n=124000*length^(-1.5) to auto-adjust the number of related hishapes in HiPath

Version 2.0.9   Dec. 17, 2012
        * debug the problem which caused project not being compiled under openSuse

Version 2.0.8   Dec. 10, 2012
        * delete --shouldermin options and calculate stable local minima using post-processing 

Version 2.0.7   Dec. 03, 2012
        * correct --minh for yielding real stable local minima

Version 2.0.6   Dec. 01, 2012
        * add option --minh

Version 2.0.5   Nov. 26, 2012
        * merge HiKinetics into package

Version 2.0.4   Nov. 26, 2012
        * add option -o for including local minima with energy barrier equal to 0 (shoulder local minima)

Version 2.0.3   Nov. 21, 2012
        * only Locally Minimal Hishreps (option -l) in output set by using filter during dynamical programming calculation      
  
Version 2.0.2   Oct. 24, 2012
        * in HiPath, add an option -r for using anchor secondary structures from specified file instead of generating them from sequence(s)

Version 2.0.1   Aug. 14, 2012
        * set Turner1999 as default energy parameters

Version 2.0.0   Aug. 3, 2012
        * start implementing HiKinetics 

Version 1.1.4   July 30, 2012
        * list all members of a specific hishape

Version 1.1.3   July 26, 2012
        * RapidHishapes

Version 1.1.2   July 18, 2012
        * a substructure B will be filtered out if its free energy is higher than another substructure (on the same subword) whose hishape is a subset or superset of hishape of B.	

Version 1.1.1   June 25, 2012
        * a substructure A will not be added in the external loop or multiloop if the free energy of A > 0 kcal/mol

Version 1.1.0   June 22, 2012
        * speed up the computation of accumulated hishape probabilities

Version 1.0.27   June 19, 2012
        * add probability calculation

Version 1.0.26   June 16, 2012
        * add microstate grammar in gap in order to generate more efficient c++ code

Version 1.0.25   June 14, 2012
        * set default value of ratio(\lambda) as 5

Version 1.0.24   June 01, 2012
        * update cpp files with new version of gapc compiler: bellmansgapc-2012.05.07

Version 1.0.23   May 29. 2012
        * remedy k-best behavior

Version 1.0.22   May 24. 2012
        * get rid of all ruby scripts
        * reconstruct the c++ code into 3 programs: RNAHeliCes, HiPath and HiTed

Version 1.0.21   Mar. 08. 2012
        * revise the discription of -m option in RNAHeliCes

Version 1.0.20   Mar. 05. 2012
        * merge functions of RNAHiTed.rb into RNAHeliCes

Version 1.0.19   Mar. 04. 2012
        * debug multiple hishape alignment with hishape type h+

Version 1.0.18   Mar. 02. 2012
        * fit in with the new version of gapc compiler (2012-02-14) 

Version 1.0.17   Mar. 01. 2012
        * get rid of hishape type limitation during aligning multiple hishapes

Version 1.0.16   Feb. 15. 2012
        * add an option for all-to-all hishrep calculation

Version 1.0.15   Jan. 12. 2012
        * fix a bug that caused reading empty lines during loading file

Version 1.0.14   Jan. 10. 2012
        * optimize the branch-and-bound strategy during calculating the minimum Sum-Of-Pairs (SP) score

Version 1.0.13   Jan. 09. 2012
        * remedy a logical mistake on calculating the minimum Sum-Of-Pairs (SP) score

Version 1.0.12   Jan. 05. 2012
        * detect the file ScoreScheme.ini automatically within project
        * restore the option -a for aligning hishapes pairwise that was commented in last version

Version 1.0.11   Jan. 03. 2012
        * add option -u for calculating the minimum Sum-Of-Pairs (SP) score of a multiple hishape alignment

Version 1.0.10   Dec. 19. 2011
        * extend the energy parameters and add a new option T for setting temperature for these energy parameters (fit the new features of gapc)

Version 1.0.9   Dec. 14. 2011
        * reconstruct project so that it links the rnalib and rtlib parts in linkage phase

Version 1.0.8   Nov. 28. 2011
        * add probability calculation as option -p
        * add a coefficient to scale the score calculated from helix index distance 

Version 1.0.7   Nov. 15. 2011
        * add hishape comparison and hishape transformation from secondary strcuture

Version 1.0.6
        * add new hishape abstract level h_plus and redefine the other level meanings:
            hishape h  => helix indices from hairpin loops are considered
            hishape h+ => helix indices from hairpin loops and nested information are considered
            hishape m  => helix indices from hairpin- and multiloops in nested form
            hishape b  => all helix indices in nested form. 

Version 1.0.5
        * delete filter option; add brackets to restrict relationship between multiloops and their children; update the README and pf_filter.hh, old version in 1.0.4 package

Version 1.0.4
        * add filter option

Version 1.0.3   Aug. 28. 2011
        * add match option: hishrep is computed according to a predefined hishape

Version 1.0.2   Aug. 7. 2011
        * minors

Version 1.0.1   Aug. 1. 2011
        * minors

Version 1.0     Jul. 28. 2011
        * initial version


src/hited.cc
#include "hited_util.hh"



int main(int argc, char **argv)
{  
  
    // ########################
    // ## options            ##
    // ########################
        /* options declaration */
    //bool window_mode = false;
    //#ifdef WINDOW_MODE
        //    bool window_mode = true;
    //#endif


    unsigned int hishape_type;
    float ratio = 5.0f;  // TODO: describe the ratio is correspondent to \lambda in the paper in option description
    unsigned int convert = 0;

    try {
      
        // idea is the option is not compulsory and it will be given a relexed code  
        //po::options_description desc("Allowed options", 120, 60);
        po::options_description desc("Allowed options");
	
	desc.add_options()
	("file,f", po::value< std::vector<std::string> >(), "Read sequence(s) from specified file (fas-format, see examples/README for format explanation)")
	("type,t", po::value<unsigned int>(&hishape_type)->default_value(2), "Specify hishape abstract type")
        ("ratio,r", po::value<float>(&ratio)->implicit_value(5.0f), "Specify weight between the score from helix index distance and score from helix index type conversion")
	("convert,c", po::value<unsigned int>(&convert)->implicit_value(1), "Convert secondary structure to hishape")
	("help,h", "Produce help message")
	("version,v", "Show version");


        /*
         * it can add at most additional 2 records, that means,
         * it is still probable that the sum of the records >= 2
         *
         * ./main --help -f aaa dddd ccc -s bbb
         * will get the result
         * input files are: aaa
         * input sequences are: dddd ccc bbb
         * positional options are: dddd ccc bbb
         */
        po::positional_options_description p;
        p.add("file", -1);


        po::variables_map vm;
        po::store(po::command_line_parser(argc, argv).
                  options(desc).positional(p).run(), vm);
        //po::store(parse_command_line(argc, argv, desc), vm);
        po::notify(vm);

        // option exclusions and implications
	conflicting_options(vm, "help", "version");
	conflicting_options(vm, "file", "sequence");
	conflicting_options(vm, "align", "convert");


        //std::cout << "vm.size()=" << vm.size() << std::endl;
        if (vm.count("help") || vm.size()==1) {
            std::cout << "Usage: HiTed [<first hishape>] [<second hishape>] [options]\n";
	    std::cout << "Examples:" << std::endl;
	    std::cout << "  ./HiTed '[27]' '[36.5m,(,27,38,)]' -r 1" << std::endl;
	    std::cout << "  ./HiTed ../examples/riboswitches.fas -t 1 -r 1" << std::endl;  // compare the 1st and 2nd, 3rd and 4nd
	    std::cout << "  ./HiTed ../examples/riboswitches.fas -c" << std::endl;            // convert every record 
            std::cout << desc; //<< '\n';
            return 0;
        }

        if (vm.count("version")) {
            std::cout << "HiTed 2.0.14 (Jan. 14, 2014)\n";  // TODO: use the C++ buildin date function 
            return 0;
        }
        

       if (vm.count("file"))
       {
	    std::vector<std::string> files = vm["file"].as< std::vector<std::string> >();
            if (!files.empty()) {
	        // ##################
	        // ## convert mode ##
	        if (convert==1 && files.size()==1) 
		{
		  	if (hishape_type < 1 || hishape_type > 4) 
			{
				std::cout << "The hishape abstract type can only be specified between 1 and 4." << std::endl;  // For multiple hishape alignment, 
				return 1;
			}

			std::string line;
			std::fstream fiInputFile;
			if (false == fiInputFile.is_open())
			{       
					// get first input file
					fiInputFile.open(files[0].c_str(), std::ios::in);
					// checks if the file was opened
					if ((false == fiInputFile) || (false == fiInputFile.is_open()))
					{
						// not opened, throws an exception
						throw "ERROR: loading input file (\"" + files[0] + "\")!";
					}
					
					std::string header = "";
					std::string seq = "";
					std::string ss = "";
					int line_no = 0;
					while ( getline(fiInputFile, line) )
					{

						if ( (line_no%3 == 0) ) {
						    if (line[0] == '>')
						    {
							// get header only
							//header = fasta.substr(1);
							header = line;
						    }
						    else
						    {
							throw "ERROR: wrong faa-format in input file (\"" + files[0] + "\")!";
						    }
						}
						else 
						{
						    if (line_no%3 == 1)
							seq = line;
						    if (line_no%3 == 2)
							ss = line;
						}


						if (header != "" && seq != "" && ss != "") {
							std::cout << header << std::endl;
							std::cout << seq << std::endl;
							std::cout << ss << std::endl;
                                                        std::cout << sSToHishape(header+"\n"+seq+"\n"+ss+"\n", hishape_type) /*<< " (hishape type = " << hishape_type << ")" */<< std::endl;
							header = "";
							seq = "";
							ss = "";
						}
						line_no ++;
					}
			}
			else
			{
				std::cout << "ERROR: unable to open input file (\"" + files[0] + "\")!\n";
			}
			
			if (true == fiInputFile.is_open())
			{
				fiInputFile.close();
			}
		}
                else
	        // ##################
	        // #### align mode ##
		if (files.size()==2) 
		{   
		        std::string strInput1,strInput2;
		        if (!files[0].empty() && files[0].at(0) == '[' &&  !files[1].empty() && files[1].at(0) == '[')
			{
				strInput1 = files[0];
				strInput2 = files[1];
				pairAlign(strInput1, strInput2, hishape_type, ratio);
			}
		}
		else
		if (files.size()==1) 
		{
			std::string line;
			std::fstream fiInputFile;
			if (false == fiInputFile.is_open())
			{       
					// get first input file
					fiInputFile.open(files[0].c_str(), std::ios::in);
					// checks if the file was opened
					if ((false == fiInputFile) || (false == fiInputFile.is_open()))
					{
						// not opened, throws an exception
						throw "ERROR: loading input file (\"" + files[0] + "\")!";
					}

					std::string header1 = "";
					std::string seq1 = "";
					std::string ss1 = "";
					std::string header2 = "";
					std::string seq2 = "";
					std::string ss2 = "";
					int line_no = 0;
					while ( getline(fiInputFile, line) )
					{

						if ( (line_no%6 == 0) ) {
						    if (line[0] == '>')
						    {
							// get header only
							//header1 = line.substr(1);
							header1 = line;
						    }
						    else
						    {
							throw "ERROR: wrong faa-format in input file (\"" + files[0] + "\")!";
						    }
						}
						else if ( (line_no%6 == 3) ) {
						    if (line[0] == '>')
						    {
							// get header only
							//header2 = line.substr(1);
							header2 = line;
						    }
						    else
						    {
							throw "ERROR: wrong faa-format in input file (\"" + files[0] + "\")!";
						    }
						}
						else 
						{
						    if (line_no%6 == 1)
							seq1 = line;
						    if (line_no%6 == 2)
							ss1 = line;
						    if (line_no%6 == 4)
							seq2 = line;
						    if (line_no%6 == 5)
							ss2 = line;
						}


						if (header1 != "" && header2 != "" && seq1 != "" && seq2 != "" && ss1 != "" && ss2 != "") {
						        std::cout << "#### " << header1 + " vs. " + header2 + " ####" << std::endl;  
                                                        pairAlign(sSToHishape(header1+"\n"+seq1+"\n"+ss1+"\n", hishape_type), sSToHishape(header2+"\n"+seq2+"\n"+ss2+"\n", hishape_type), hishape_type, ratio);
							header1 = "";
							header2 = "";
							seq1 = "";
							seq2 = "";
							ss1 = "";
							ss2 = "";
						}
						line_no ++;
					}
			}
			else
			{
				std::cout << "ERROR: unable to open input file (\"" + files[0] + "\")!\n";
			}
			
			if (true == fiInputFile.is_open())
			{
				fiInputFile.close();
			}
		}
		else
		{
		    std::cout << "Status: files.size()=" << files.size() << std::endl;
		}
	    }
        }
    } catch (std::exception &e) {
        std::cerr << "Exception: " << e.what() << '\n';
        std::exit(1);
    }


  return 0;
}
